### ISSUE TITLE
Are we forgetting `docker secret update` command in cli?

### ISSUE THREAD

**Issue by realityone** [docker/docker#29882]
Hi there,

I'm trying to use docker secret to manage our private keys, but I find that we support `/secrets/<id>/update` in API,  but no `docker secret update` command in cli, even in client.

- https://github.com/docker/docker/blob/master/client/interface.go#L164
- https://github.com/docker/docker/blob/master/api/server/router/swarm/cluster.go#L50

Are we just forgetting it or it's by designed? 

I'm working on Docker 1.13.0-rc4.

**Comment by realityone** [comment]
Oh, this is by designed at https://github.com/docker/docker/blob/master/daemon/cluster/secrets.go#L109.

**Comment by thaJeztah** [comment]
I think updating is currently an API only feature; the secret (data) itself is immutable, and only "labels" can be updated.  We decided that having a `secret update` command would be more confusing than "not" having this, so (at least for now) updating through the CLI is not implemented.

Thanks for checking though, and paying attention :smile: 👍 !

**Comment by s4s0l** [comment]
Hmm, so in order to update secret the only available procedure is to remove all services using it, recreate secret and start services from scratch?

**Comment by thaJeztah** [comment]
You cannot update the secret itself, but you can update the service that's using it, and add a different secret;

```
echo foo | docker secret create secret-1 -

docker service create \
  --secret src=secret-1,target=db-password \
  --name web \
  nginx:alpine

echo foo2 | docker secret create secret-2 -

docker service update \
  --secret-rm secret-1 \
  --secret-add src=secret-2,target=db-password \
  web
```

**Comment by s4s0l** [comment]
Reasonable as not many images will support password change on the fly. Thanks.

**Comment by thaJeztah** [comment]
One of the considerations to make the secret itself immutable is to allow rolling back a service definition to a previous version. A PoC version of the feature had built-in versioning of the secret itself, but this added a lot of complexity with not a lot to gain, so if was decided to keep this out.

**Comment by kleptog** [comment]
TBH, when rolling back to a previous service definition it doesn't seem at all unreasonable to get the new version of the secret. After all, the secret (generally) gives access to a resource outside the swarm, and that is not being rolled back.

Not having an update command is also annoying when you are deploying your infrastructure via ansible. It means that password rollover will never happen because you can't reliably script everything that needs to be updated.

Interestingly, "docker stack" does have all the necessary info to rebuild the system, but it would be weird if docker stack changed the name of the secret for you.

This is one of those situations where I wish docker would give you just a bit more rope.

**Comment by RehanSaeed** [comment]
> You cannot update the secret itself, but you can update the service that's using it, and add a different secret;

That seems suboptimal. I expected to see an update command too. It would be nice if the secret was updated in running containers too but I'd settle for having to redploy a service if I could get my hands on this feature.

Can we consider re-opening this issue?

**Comment by vizv** [comment]
And it's quite annoying change compose file every time when updating a config or secret.

**Comment by MLescaudron** [comment]
I don't understand why the command is unavailable because of UpdatedAt field : 
```
[
    {
        "ID": "l2qnmoxtlopsgij9ayu4pgbsj",
        "Version": {
            "Index": 47914
        },
        "CreatedAt": "2017-09-20T03:46:16.035020938Z",
        "UpdatedAt": "2017-09-20T03:46:16.035020938Z",
        "Spec": {
            "Name": "test_1",
            "Labels": {}
        }
    }
]
```

So, while waiting, I created a script that allows to do it easily.
It updates all services using the secret.
If this interests you : https://gist.github.com/MLescaudron/e8248d32d3a5b8caaf622c1a829cf067

**Comment by rgardam** [comment]
Also this is quite painful as you cant use variables in the docker-compose file for secret versions.

**Comment by stevelacy** [comment]
As a workaround for using `docker stack deploy` (which would need to be modified for the new secret name) I am using this script to update a swarm wide secret.
It does have the issue of two `docker service update` rolls.

https://github.com/stevelacy/docker-secret-update

**Comment by ilyasotkov** [comment]
Follow this issue <https://github.com/docker/cli/issues/667> and pull request <https://github.com/docker/cli/pull/668>, which will add a "name" field to both configs and secrets. It will allow updates via environmental variables like so:
```yaml
version: "3.5" # future Compose file version

secrets:
  my-secret:
    name: my-secret-${VERSION} # name of the Docker Secret with variable interpolation
    file: ./secret.html

configs:
  my-config:
    name: my-config-${VERSION} # same thing for configs
    file: ./default.conf

services:
  web:
    image: nginx:1.13.6-alpine
    ports:
      - "80:80"
    secrets:
      - source: my-secret # reference by the name of the key, consistent with how it is for volumes now in 3.4
        target: /usr/share/nginx/html/index.html
    configs:
      - source: my-config # same thing for configs
        target: /etc/nginx/conf.d/default.conf
```

**Comment by yangm97** [comment]
So, the solution is adding arbitrary names to secrets? Why not do it like environment variables management? How do I remove orphaned secrets?

**Comment by immerrr** [comment]
> Updating is currently an API only feature; the secret (data) itself is immutable, and only "labels" can be updated.

Ok, but the problem is that the labels cannot be updated without hipping through the hoops of finding all services that use that specific secret label, unlinking them manually and then doing the `rm; create`. 

Making secrets immutable is a nice feature, so why don't do it the Git way? I mean, you already do it with images and their hashes, where each image is uniquely identified with a hash, but you can also add a human-readable tag and then it will be resolved upon container creation. Then you are free to push a new image and update the tag, but the old image is still there referenceable by its hash until it's garbage-collected.  @thaJeztah, was there a reason why the same approach was unapplicable for secrets?

**Comment by thaJeztah** [comment]
@immerrr see some of my comments in https://github.com/moby/moby/issues/35048#issuecomment-333778750 (with a possible approach)

**Comment by dlibanori** [comment]
It is very frustrating to manage secrets this way :(

I would love to have my secrets versioned, just update it and docker relaunches my services, just rollback it and docker gets my previous secret.

**Comment by timrazik** [comment]
I know this is an older thread but as someone who manages a large number of secrets, I have a suggestion here.  If you want to change a secret with minimal to no impact you can create credentials that work in parallel. This comes in handy if you want to expire credentials periodically to meet security compliance or close gaps in cases where people change teams or leave your organization.  Let's take an example of a database ID and Password.  What we can do is actually create TWO sets of credentials with the same security permissions.  

```
database:  myDockerDB
userId_A:    appUserID_A
password_A:  some_password#1234
userId_B:  appUserID_B     // This User would be INACTIVE for now.
password_B:  some_other_password#9876

```
In our docker stack/compose file we would then map our secrets as such

```
services:
  myService :
    image: dtr.docker.com/mycompany/dockerService:1.0.0
    environment:
      - JAVA_OPTS=-Xms512m -Xmx512m -XX:+UseG1GC -XX:+UseStringDeduplication -XX:-TieredCompilation -XX:+ParallelRefProcEnabled
    deploy:
      mode: global
    networks:
      - bridge
    configs:
    secrets:
      - mode: 444
        source: userId_A
        target: userId
      - mode: 444
        source: password_A
        target: password

```

Now here comes the fun part.  When we want to change the password, what you do is have your DBA ACTIVATE userId_B and you set your password for it.  BOTH userId_A and userId_B would be active for an overlap (say 7 days).

You would then just update your stack file or service update to map the B credentials.  This would be allowed because no service is utilizing B.  The beauty here is that no code change is required as you are not changing the key that the code is using to look up the credentials. Now you can roll through your service(s) during the overlap period.  Once complete you have your administrator deactivate the old account so that it can no longer be used.

```
services:
  myService :
    image: dtr.docker.com/mycompany/dockerService:1.0.0
    environment:
      - JAVA_OPTS=-Xms512m -Xmx512m -XX:+UseG1GC -XX:+UseStringDeduplication -XX:-TieredCompilation -XX:+ParallelRefProcEnabled
    deploy:
      mode: global
    networks:
      - bridge
    configs:
    secrets:
      - mode: 444
        source: userId_B
        target: userId
      - mode: 444
        source: password_B
        target: password

```

**Comment by thaJeztah** [comment]
@Sammons what version of Docker are you running? I tried to reproduce that situation, but wasn't able to;

```bash
$ printf "secret1" | docker secret create secret1 -
$ printf "secret2" | docker secret create secret2 -
```

Create a service using secret1

```bash
$ docker service create --secret source=secret1,target=/somewhere --name myservice nginx:alpine
7mamd13lh282p5d5i423lmabd
overall progress: 1 out of 1 tasks 
1/1: running   [==================================================>] 
verify: Service converged 
```

Update the service, and replace `secret1` with `secret2`

```bash
$ docker service update --secret-rm secret1 --secret-add source=secret2,target=/somewhere myservice
myservice
overall progress: 1 out of 1 tasks 
1/1: running   [==================================================>] 
verify: Service converged 
```

If you have ways to reproduce on a current version of Docker, could you open an issue for that?

**Comment by Sammons** [comment]
@thaJeztah I will try updating to 18.* and get back to you. I know these nodes are running 17-something.

**Comment by thaJeztah** [comment]
@Sammons it could be a client-only change, so for a quick workaround, you could just try with a more recent version of the Docker CLI

**Comment by Sammons** [comment]
@thaJeztah for reference this situation happened with client version: 18.05.0-ce and with server version 17.12.0-ce; I am trying to update the server to fix.

**Comment by Sammons** [comment]
@thaJeztah Got it. The problem arises when I try removing a secret that doesn't exist, so the error is correct. The error thrown is just not pointing to my root problem (that I'm removing a secret that does not exist).

**Comment by thaJeztah** [comment]
Ah; looks like something to improve: could you open a new issue for that with minimal reproduction steps?

**Comment by markvr** [comment]
+1 for implementing a simpler way to update a secret please!   Currently we store compose files in version control, with encrypted passwords as environment variables, and a simple script on Jenkins that decrypts them before deploying.  This means the passwords are in plaintext in the services.

We'd like to remove this and move to secrets but the whole theatre around secrets is just too complicated.   Trying to maintain secrets in a CI pipeline requires a lot of custom work to encrypt them in version control, generate a unique name for the secret,  deploy the secret before the service, then update the compose file to have the new unique name for the secret (or keep track of an external variable name), deploy the compose file, and finally remove the old secret.  What a faff!

My ideal situation would be a public/private key pair, which is used to encrypt the secret.  The compose file then looks like:

`secrets:
   my-secret: ENC(......)`

This can safely be version controlled, and when deployed Docker then decrypts the secret with the private key, and updates any services that were using it.  This means  the encrypted value of the secret is treated the same as any other configuration value, and would massively simplify getting secrets out of environment variables.

I realise a lot of people probably externalise their secret management using something like Hashicorp Vault or similar, but that is a whole extra product, and it would be really cool if this could just be kept simple within Docker.

@thaJeztah - Could this issue be reopened, or is it worth me creating a new issue with this feature request?

**Comment by jamiejackson** [comment]
Adding my script to the list: https://gist.github.com/jamiejackson/a1818acedaeb9c3cd70bafac86a0100b

As far as I can tell, it might be an improvement over @stevelacy's (which uses another file and needs two service updates). I just finished it, though, so I'm not sure what its flaws will be.

**Comment by BluSyn** [comment]
I am frustrated by the need for complicated solutions to a simple problem. Using docker-compose + secrets w/ `docker stack deploy` is essentially impossible with the current paradigm.

It should not be required to manually update the docker-compose file every time I update the secret. Currently not changing the secret name in compose file causes `docker stack deploy` to fail if any secret file has been changed. This is frustrating since this would not happen in any other context: if anything else in the docker-compose file has changed, even external files such as an `env` file, docker automatically updates the services as one would expect. The current behavior makes auto-deployments + secret management essentially impossible while still using version controlled configuration.

Sharing the script I am currently using to solve this problem:
https://gist.github.com/BluSyn/71a2040d610e06bf8ee06a9480d17fd3

This script will automatically find any service(s) on the swarm using the secret with the specified file name and update them with a new secret. Should be run on a manager node.

This solution requires direct parsing of both the `docker-compose.yml` file and the current state of the swarm. The advantage is future runs of `docker stack deploy` will succeed. It will have a side-effect of re-creating secrets + re-deploying all services manually updated since the last full deployment. One nice thing of this script is the limited information an admin needs to give in order to update a secret and doesn't require a full stack deploy.

In my opinion this is exactly how `docker stack deploy` should handle secret file updates. This would allow secrets to maintain their "immutability", but the expected behavior of deploying with a compose file is maintained.

**Comment by iFreeon** [comment]
I'm using docker for the first time and also puzzled why I can't update secrets.  I'm deploying stacks using docker-compose. 

How do rolling updates work if I have to take the whole stack down to remove and re-add a secret?

**Comment by kennell** [comment]
I really is silly that this is not supported. Consider a web application where API keys for third-party services are rotated on a regular basis etc. -- tons of use cases where it would be useful to update a secret.

**Comment by val-roam** [comment]
we store our ssl certs and keys in docker secrets, when its time to update those - its a massive pain in the ass! have to go hunt down everything that uses these certs (which is..everything) change the secret name to _v2 or _v3 or whatever, redeploy, argh ...

and that's just ssl certs. there're plenty of secrets that need to be rotated on regular basis, and much more often than ssl certs!

what kind of work is being done around solving this situation?

**Comment by er1z** [comment]
Reopen, please!

**Comment by DD202** [comment]
A pretty straight forward solution is to calculate the hash of the file being used as the secret and inject that as the name of the secret. This way if the secret changes, a new one will be created during the docker stack deploy process. A process is then needed to clean up orphaned secrets long term, but it eliminates a lot of the headache from a dev standpoint.

**Comment by jamiejackson** [comment]
@DD202 You and I must have different definitions of "straightforward," but that's exactly what I'm doing.

Although I have this going in production, it wasn't quite ready to be untangled and put into a repo for public consumption. However, I've just posted it in its raw and ugly state (and maybe I'll get around to making it more presentable some day).

https://github.com/jamiejackson/docker-secret-hash

**Comment by Xanatosinc** [comment]
I landed here trying to Google my way through this. This approach could still maybe be improved upon, but there's a one-line command for rotating secrets about halfway down the page here (two lines if you include creating the new secret), under Rotating Docker Secrets: https://semaphoreci.com/community/tutorials/managing-secrets-in-docker-swarm

Suppose you have an existing service `service_app` with a connected secret `secret_1` which is mapped to `/path/in/container`:
```
docker secret create secret_2 /updated/file/in/host
docker service update --secret-rm secret_1 --secret-add source=secret_2,target=/path/in/container service_app
```

**Comment by r0039** [comment]
In fact, docker secret should be managed like Git way or version control. It is really useful to to manage, scale the applications. 

I give an example: 
First day at new company, the existing SSL has just expired. So I have to update the docker secret with the new SSL. On docker manager, all of docker containers are using the same docker-secret-name. So better way to backup the current docker-secret-key, then update the existing docker-secret with a new SSL. After that, I only need to re-run docker service to apply the change. 

That's save time. Thank you and hope you consider to code this feature.

**Comment by bdw429s** [comment]
@thaJeztah Why are all the comments on this thread being ignored?  The current behavior is terrible for usability and need to be revisited.  Updating a secret value and having it automatically picked up by any services using it without needing to remove the secret or touch the services is a real and common scenario.

**Comment by er1z** [comment]
AFAIR even K8s has secrets immutable, so IMO it's time to learn new approach.

**Comment by immerrr** [comment]
> AFAIR even K8s has secrets immutable

Secrets are mutable on K8S. There has been a rather recent addition to be able mark secrets and configmaps as immutable, so that k8s doesn't have to spend resource checking for updates, but that's optional.

**Comment by agostlg** [comment]
Mutable secrets please! Would this even be considered?

**Comment by agostlg** [comment]
Mutable secrets please! Would this even be considered?

**Comment by mesudip** [comment]
Ok fine, If versioning in secret is too complex, At least we can have command to make it easier to do the `secret-rm` and `secret-add` ?
```bash
# basically finds all the services using old_secret and replace them with new_secret
docker service  replace-secret old_secret   new_secret
```
or may be a shorthand
```
echo -n "newPassword" | docker secret create --replaces=old_secret new_secret -
```

**Comment by souchprod** [comment]
Reopen please, current solution is just not good enough

**Comment by Tillman32** [comment]
How do we manage updating secrets via CI/CD without an update command?

**Comment by tjdraper** [comment]
Just chiming in here on how absolutely awful the experience is that we can't update secrets. Secret update/rotation as currently implemented requires WAY too many hoops.

In our deploy pipeline we can't simply create/update a secret from the pipeline's secrets. All the current solutions are pretty intense. And simply making secrets mutable would solve every single problem in an extremely elegant way for a very, very, VERY common use case.

**Comment by Radiergummi** [comment]
@thaJeztah  So, what am I supposed to do to rotate a secret in a CI pipeline? I have to start by creating a docker-compose configuration with a version in the secret name, perhaps from the current commit hash (`my-secret-${COMMIT}`). That requires the [awkward workaround](https://github.com/docker/cli/issues/939#issuecomment-493920237) of doing `--compose-file <(docker-compose --file docker-compose.yaml config)` to expand environment variables, which works with `docker-compose`, but not `docker compose`. It's an incredible mess.

This works, but litters my cluster with old secrets, so I need to clean those up. But how would I know the previous commit hash in a CI pipeline? So instead, I'll have to iterate over all secrets in the cluster with the same base name, and remove those. That needs to be done after the actual deployment is complete though, because the secret'll be in use otherwise. That, in turn, means I need to wait for the old containers to shut down.  
All in all, we're looking at a _giant_ shell script to update the value of my secrets. How did this pass the reality check? Are secrets supposed to only work for setup-and-forget deployments?

Immutable secrets are a great idea in principle. But they're utterly useless without versioning and an internal, versioned identifier - WHY do secrets determine uniqueness via their human-readable name..? All of this would be so much less pain if secrets got a random hash, just like other docker resources.

**Comment by Radiergummi** [comment]
Funny enough, the awkward workaround doesn't even work anymore because `docker stack` and `docker compose` have different opinions on how a docker-compose.yaml file should look like, as documented here: https://github.com/docker/cli/issues/2527#issuecomment-1197727199

Instead of the following, already-awkward version:
```bash
docker stack deploy --compose-file <(docker-compose --file docker-compose.yaml config)
```

We have to use the following, even more convoluted one:
```bash
cat (echo -e "version: '3.8'"; docker-compose --file docker-compose.yaml config | sed '/^name: build$/d') | docker stack deploy --compose-file -
```

These issues span so many tools and repositories, I don't even know where to properly report them. And even if I did, all existing issues on similar topics go without resolution for years... The situation is beyond frustrating.

**Comment by s4ke** [comment]
For us it turns out to get the secret and config rotation going in a simple way, augmenting the stack file with a wrapper to `docker stack deploy` was the most convenient solution.

We store our secrets encrypted in git and decrypt them before running the deployment. Then our wrapper hashes the contents of secrets and configs and uses the first 12 characters of that sha sum to generate a new secret name.

https://github.com/neuroforgede/docker-stack-deploy/

All of this is done automatically without any environment name expansion. The python script understands enough of the yaml spec for stacks so that it actively searches for all occurences of these secrets/configs and uses them.

How do we clean up the extra secrets?

Right now we dont have it set up as we dont have a big issue with the auto rotating secrets, but in the future, we can automate this with a cronjob running on every node to clean up all secrets older than X amount of time. If the secret is older than X but is still in use by a service it will not be deleted.

**Comment by Radiergummi** [comment]
> We store our secrets encrypted in git and decrypt them before running the deployment. Then our wrapper hashes the contents of secrets and configs and uses the first 12 characters of that sha sum to generate a new secret name.

I don't know how you implemented this, but it sounds dangerous: If you store your secrets in the repository, there is no way for you to rotate the encryption secret securely. The values encrypted with a potentially leaked secret will still be contained within your repository and can be easily retrieved by an attacker with access to the repository (even just a local copy, for example on the device of a grumpy ex-employee).  
I hope you know what you are doing.

**Comment by s4ke** [comment]
It is not optimal, especially having the local copies part, I give you that. The next iteration of this tooling will probably use something like vault to store the secrets before injecting them instead of storing data on disk.

About encryption: Data is encrypted with the public key of every relevant user. Secret revocation is then "just" the removal of a public key from a recipients file + rotation of all relevant secrets.

In any case, I don't think this is the place to discuss this.

**Comment by ikegentz** [comment]
New Year, same problem. It would be great if this functioned like K8s secrets

**Comment by ohmydevops** [comment]
New Year, same problem. It would be great if this functioned like K8s secrets

**Comment by shaunlowis** [comment]
Another year, yet this seemingly basic functionality is still not available. Please re-open, unless I have missed something?

**Comment by nikelborm** [comment]
@shaunlowis wasn't it stated previously that secrets are immutable by design?
